File: src/main/java/io/jenkins/plugins/gamekins/challenge/CoverageChallenge.java
package io.jenkins.plugins.gamekins.challenge;

import hudson.FilePath;
import io.jenkins.plugins.gamekins.util.JacocoUtil;
import org.jsoup.nodes.Document;

import java.io.IOException;

public abstract class CoverageChallenge implements Challenge {

    final JacocoUtil.ClassDetails classDetails;
    final int fullyCoveredLines;
    final int partiallyCoveredLines;
    final int notCoveredLines;
    final double coverage;
    double solvedCoverage = 0;
    final String branch;
    final long created = System.currentTimeMillis();
    long solved = 0;

    public CoverageChallenge(JacocoUtil.ClassDetails classDetails, String branch, FilePath workspace)
            throws IOException, InterruptedException {
        this.classDetails = classDetails;
        this.branch = branch;
        Document document = JacocoUtil.generateDocument(JacocoUtil.calculateCurrentFilePath(workspace,
                this.classDetails.getJacocoSourceFile(), classDetails.getWorkspace()));
        this.fullyCoveredLines = JacocoUtil.calculateCoveredLines(document, "fc");
        this.partiallyCoveredLines = JacocoUtil.calculateCoveredLines(document, "pc");
        this.notCoveredLines = JacocoUtil.calculateCoveredLines(document, "nc");
        this.coverage = classDetails.getCoverage();
    }

    @Override
    public long getCreated() {
        return this.created;
    }

    @Override
    public long getSolved() {
        return this.solved;
    }

    abstract String getName();

    @Override
    public String printToXML(String reason, String indentation) {
        String print = indentation + "<" + getName() + " created=\"" + this.created + "\" solved=\"" + this.solved
                + "\" class=\"" + classDetails.getClassName() + "\" coverage=\"" + this.coverage
                + "\" coverageAtSolved=\"" + this.solvedCoverage;
        if (!reason.isEmpty()) {
            print += "\" reason=\"" + reason;
        }
        print += "\"/>";
        return print;
    }
}

File: src/main/kotlin/io/jenkins/plugins/gamekins/challenge/CoverageChallenge.kt
package io.jenkins.plugins.gamekins.challenge

import hudson.FilePath
import io.jenkins.plugins.gamekins.util.JacocoUtil.ClassDetails
import io.jenkins.plugins.gamekins.util.JacocoUtil.calculateCoveredLines
import io.jenkins.plugins.gamekins.util.JacocoUtil.calculateCurrentFilePath
import io.jenkins.plugins.gamekins.util.JacocoUtil.generateDocument

/**
 * Abstract class to generate basic information about the class used for generating a [CoverageChallenge].
 *
 * @author Philipp Straubinger
 * @since 1.0
 */
abstract class CoverageChallenge(val classDetails: ClassDetails, val branch: String, workspace: FilePath?)
    : Challenge {

    val coverage: Double
    val fullyCoveredLines: Int
    val notCoveredLines: Int
    val partiallyCoveredLines: Int
    var solvedCoverage = 0.0
    //TODO: Fix
    @get:JvmName("getCreated_") val created = System.currentTimeMillis()
    @get:JvmName("getSolved_") var solved: Long = 0

    /**
     * Calculates the number of fully, partially and not covered lines, and the coverage of the class itself.
     */
    init {
        val document = generateDocument(calculateCurrentFilePath(workspace!!,
                classDetails.jacocoSourceFile, classDetails.workspace))
        fullyCoveredLines = calculateCoveredLines(document, "fc")
        partiallyCoveredLines = calculateCoveredLines(document, "pc")
        notCoveredLines = calculateCoveredLines(document, "nc")
        coverage = classDetails.coverage
    }

    override fun getCreated(): Long {
        return created
    }

    /**
     * Returns the name of the class of the current [CoverageChallenge].
     */
    abstract fun getName(): String

    override fun getSolved(): Long {
        return solved
    }


    override fun printToXML(reason: String, indentation: String): String {
        var print = (indentation + "<" + getName() + " created=\"" + created + "\" solved=\"" + solved
                + "\" class=\"" + classDetails.className + "\" coverage=\"" + coverage
                + "\" coverageAtSolved=\"" + solvedCoverage)
        if (reason.isNotEmpty()) {
            print += "\" reason=\"$reason"
        }
        print += "\"/>"
        return print
    }
}
